[lldb/formatter] Add Swift.UnsafeBufferPointer data formatter

medismailben opened the pull request:
This patch adds support for Swift's UnsafeBufferPointer and
UnsafeMutableBufferPointer data formatting.

It introduces a summary provider and a synthetic frontend creator for
these types.

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

fredriss commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:
```diff
@@ -0,0 +1,37 @@
+func main() {
+  // UnsafeBufferPointer
+  let arr = [ 1, 2, 3 ]
+  let buf = arr.withUnsafeBufferPointer { $0 }
```
This is wrong. The `UnsafeBufferPointer` you get inside the closure is valid only there, returning it is undefined. You need to either examine the value while inside the closure, or create the `Unsafe*Pointer` differently.

fredriss commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:
```diff
@@ -0,0 +1,37 @@
+  // UnsafeRawBufferPointer
+  let rawbuf = UnsafeRawBufferPointer(buf)
+  // UnsafeRawMutablePointer
+  let mutrawbuf = UnsafeMutableRawBufferPointer(mutbuf)
```
Looks like those variables are defined but not tested?

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,257 @@
+
+  auto &item = m_children[idx];
+  if (!item.second) {
+    DataBufferSP buffer_sp(new DataBufferHeap(m_ptr_size, 0));
```
Is there no cheaper way to do this? I think `DataBufferHeap` is going to do a malloc. It's a waste for a temporary of maximum 8 bytes.

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,257 @@
+    } break;
+    default:
+      lldbassert(false && "pointer size is not 4 nor 8");
+    }
```
Isn't this whole thing equivalent to `memcpy(buffer_sp->GetBytes(), &item.first, m_ptr_size)` ?

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,257 @@
+    UnsafeBufferPointerSyntheticFrontEnd::GetChildAtIndex(size_t idx) {
+  if (!m_unsafe_ptr.IsValid() && idx > m_unsafe_ptr.GetCount())
+    return ValueObjectSP();
+
+  addr_t m_start_addr = m_unsafe_ptr.GetStartAddress();
+  const uint32_t num_children = CalculateNumChildren();
+
+  if (idx >= num_children)
+    return lldb::ValueObjectSP();
+
+  if (m_children.empty()) {
+    ProcessSP process_sp = m_exe_ctx_ref.GetProcessSP();
+    if (!process_sp)
+      return lldb::ValueObjectSP();
+
+    Status error;
+    addr_t value_at_idx = 0;
+
+    uint32_t tries = 0;
+    uint32_t test_idx = 0;
+
+    // Iterate over inferior memory, reading key/value pointers by shifting each
+    // cursor by test_index * m_ptr_size. Returns an empty ValueObject if a read
+    // fails, otherwise, continue until the number of tries matches the number
+    // of childen.
+    while (tries < num_children) {
+      value_at_idx = m_start_addr + (test_idx * m_ptr_size);
+      value_at_idx = process_sp->ReadPointerFromMemory(value_at_idx, error);
+      if (error.Fail())
+        return lldb::ValueObjectSP();
+
+      test_idx++;
+
+      if (!value_at_idx)
+        continue;
+      tries++;
+
+      m_children.push_back(std::make_pair(value_at_idx, nullptr));
+    }
+  }
```
Can this be a `for` loop? Why is a null pointer just skipped? It would create am inconsistency between the indices in memory and in the values we display which would be a huge issue.

But more generally this seems wrong? It's saying that all the values in a buffer are pointer-sized, which I'm pretty sure is wrong. The UnsafeBuffer elements are like Array elements. They are the size of their underlying type I would guess.

dcci replied:
My main concern with this function is that it really feels a cut'n'paste (and probably is) of other formatters. I looked at `NSDictionary` yesterday, and it contains the same code (I think the variables are also named similarly, e.g. `tries`). I wonder whether we can have a general solution and refactor all these formatters to use a single utility that walks memory and produces a generic pair.

fredriss commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:
```diff
@@ -0,0 +1,37 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
+struct HasInt {
+  var value: Int
+  
+  init(_ x: Int) {
+    value = x
+  }
+}
```
This struct is unused?

You need to test with more data types. Bigger structs, classes, existentials, enums, ...

fredriss commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/TestSwiftUnsafeTypeFormatters.py:
```diff
@@ -0,0 +1,63 @@
+            lldbutil.run_to_source_breakpoint(
+                self, 'break here', lldb.SBFileSpec('main.swift'))
+
+#        def cleanup():
+#            self.runCmd(
+#                "settings set target.max-children-count 256",
+#                check=False)
+
+#        self.addTearDownHook(cleanup)
```
Please don't add commented ou code.

dcci replied:
+1

fredriss commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/TestSwiftUnsafeTypeFormatters.py:
```diff
@@ -0,0 +1,63 @@
+# TestSwiftTypeAliasFormatters.py
```
Wrong title.

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,257 @@
+    return false;
+
+  ValueObjectSP child_valobj_sp = nullptr;
+  ValueObjectSP synthetic = value_sp->GetSyntheticValue();
+  if (synthetic)
+    child_valobj_sp = synthetic->GetChildAtIndex(0, true);
+  if (!child_valobj_sp)
+    child_valobj_sp = value_sp->GetChildAtIndex(0, true);
+  if (!child_valobj_sp)
+    return false;
```
I think this logic deserves comments to explain which cases will lead to which control flow and corresponding testing coverage.

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,257 @@
+  if (!value_sp || !value_sp->GetNumChildren())
+    return false;
+
+  ValueObjectSP some_value_sp = value_sp->GetChildAtIndex(0, true);
```
Pleas use more descriptive names. `value_sp`, `child_value_sp` and `some_value_sp` are way too generic and sometimes misleading. It took me a while to figure out that `child_value_sp` was not about a child of the buffer, but just a child of the original `value_sp`.

dcci commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/TestSwiftUnsafeTypeFormatters.py:
```diff
@@ -0,0 +1,63 @@
+                    ])
+
+
+if __name__ == '__main__':
+    import atexit
+    lldb.SBDebugger.Initialize()
+    atexit.register(lambda: lldb.SBDebugger.Terminate())
+    unittest2.main()
```
Unneeded.

dcci commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:
```diff
@@ -0,0 +1,37 @@
+  let rawbuf = UnsafeRawBufferPointer(buf)
+  // UnsafeRawMutablePointer
+  let mutrawbuf = UnsafeMutableRawBufferPointer(mutbuf)
+
+  print("break here")
```
This feels like a good candidate for a lldbinline test.

dcci commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/TestSwiftUnsafeTypeFormatters.py:
```diff
@@ -0,0 +1,63 @@
+# TestSwiftTypeAliasFormatters.py
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
```
I remember 2016 fondly. For new tests, we can probably drop the copyright given they'll go stale.

dcci commented on lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp:
```diff
@@ -270,6 +270,30 @@ static void LoadSwiftFormatters(lldb::TypeCategoryImplSP swift_category_sp) {
       "Swift.UnsafeMutablePointer",
       ConstString("^Swift.UnsafeMutablePointer<.+>$"), basic_synth_flags, true);
 
+  AddCXXSummary(
+      swift_category_sp,
+      lldb_private::formatters::swift::UnsafeBufferPointerSummaryProvider,
+      "Swift.UnsafeBufferPointer",
+      ConstString("^Swift.UnsafeBufferPointer<.+>$"), summary_flags, true);
+  AddCXXSynthetic(swift_category_sp,
+                  lldb_private::formatters::swift::
+                      UnsafeBufferPointerSyntheticFrontEndCreator,
+                  "Swift.UnsafeBufferPointer",
+                  ConstString("^Swift.UnsafeBufferPointer<.+>$"),
+                  basic_synth_flags, true);
+  AddCXXSummary(
+      swift_category_sp,
+      lldb_private::formatters::swift::UnsafeBufferPointerSummaryProvider,
+      "Swift.UnsafeMutableBufferPointer",
+      ConstString("^Swift.UnsafeMutableBufferPointer<.+>$"), summary_flags,
+      true);
+  AddCXXSynthetic(swift_category_sp,
+                  lldb_private::formatters::swift::
+                      UnsafeBufferPointerSyntheticFrontEndCreator,
+                  "Swift.UnsafeMutableBufferPointer",
+                  ConstString("^Swift.UnsafeMutableBufferPointer<.+>$"),
+                  basic_synth_flags, true);
+
```
I wonder if you can use some regex magic and reduce the duplication here given they both dispatch the same version.

medismailben replied:
We could do that but that would make the description less accurate, for instance, If I registered a single formatter for both the Immutable and Mutable using regexes, they would both show up as "Swift.UnsafeBufferPointer". I'm not sure how important is that description.

fredriss replied:
does the description matter though? We could change the description string to be `Swift.Unsafe(|Mutable)BufferPointer` if it's only used for displaying in `type summary`

dcci replied:
The description doesn't quite matter because, as pointed out by Fred, it's used only on `type summary`. The latter is only used to understand which formatter actually triggered, so, in this case, given the formatter triggering is exactly the same, it won't matter. My $0.02.

dcci reviewed (commented):
Thanks for taking care of this. While I think the logic is mostly correct, this opens the opportunity to do some things better. I outlined some of my concerns inline.

medismailben commented:
I updated the patch to address @dcci and @fredriss comments:
- I reimplemented the logic to read the entire contiguous buffer once (instead of doing a read per entry). I also added support for complex generic types like `struct` and `enum` and added comments and documentation to explain the 'value-providing synthetic children' workaround and clarify the `UnsafeBufferPointer` structure layout.
- As suggested, I unified the mutable and immutable data-formatter registration.
- Finally, I rewrote the test to prevent the UB, did some cleanup and added more test cases.

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,249 @@
+      return false;
+
+    for (size_t i = 0; i < num_children; i++) {
+      DataBufferSP buffer_sp(new DataBufferHeap(
+          m_buffer_sp->GetBytes() + i * m_element_stride,
+          m_element_stride)); // Create a buffer from contiguous memory
```
Do not create a new DataBufferHeap for each child. You can create a DataExtractor from the main DataBuffer with an offset.

dcci replied:
Reuse the `DataBufferHeap`

dcci commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/TestSwiftUnsafeTypeFormatters.py:
```diff
@@ -0,0 +1,18 @@
+# TestSwiftUnsafeTypeFormatters.py
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See https://swift.org/LICENSE.txt for license information
+# See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ------------------------------------------------------------------------------
+"""
```
Doesn't need copyright.

dcci commented on lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:
```diff
@@ -0,0 +1,59 @@
+// main.swift
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
```
+1

dcci commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,249 @@
+  if (valobj_sp)
+    Update();
+
+  if (!m_unsafe_ptr.IsValid())
+    return;
+}
```
This is a very weird idiom. It feels like dead code?

dcci commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,249 @@
+  if (!m_valobj.GetNumChildren())
+    return false;
+
+  //  ▿ UnsafeBufferPointer
+  //    ▿ _position : Optional<UnsafePointer<Int>>
+  //      ▿ some : UnsafePointer<Int>
+  //        - pointerValue : Int
+  //    - count : Int
```
Please expand this comment.

dcci commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,249 @@
+    : m_valobj(*valobj.GetNonSyntheticValue().get()) {}
+
+bool SwiftUnsafeBufferPointer::IsValid() const {
+  return m_count != UINT64_MAX && m_start_addr != LLDB_INVALID_ADDRESS &&
+         m_start_addr != 0;
```
Why do you need this?

medismailben committed:
[lldb/formatter] Add Swift.UnsafeBufferPointer data formatter

This patch adds support for Swift's UnsafeBufferPointer and
UnsafeMutableBufferPointer data formatting.

It introduces a summary provider and a synthetic frontend creator for
these types.

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

medismailben commented:
Added more tests (value associated enums, generic class, existential type (archetypes) ...)
Addressed recent feedbacks.
Removed copyright header in tests.

medismailben commented:
The macOS failure happened on a Swift test. Given that I only touched to LLDB, I assume this is a flaky test.

medismailben commented:
@swift-ci test macOS Platform

medismailben commented:
https://github.com/apple/swift/pull/32152 Should fix the test failure. Let's try re-running the macOS CI again!

medismailben commented:
@swift-ci test macOS Platform

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,250 @@
+  if (!value_provided_child_sp)
+    return false;
+
+  m_count = value_provided_child_sp->GetValueAsUnsigned(UINT64_MAX);
```
I don't know under which conditions this can fail, but the API makes it look like it can. I would add a defensive check for UINT64_MAX after this.

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,250 @@
+  if (!pointer_value_sp)
+    return false;
+
+  m_start_addr = pointer_value_sp->GetValueAsUnsigned(LLDB_INVALID_ADDRESS);
```
ditto

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,250 @@
+    : SyntheticChildrenFrontEnd(*valobj_sp.get()),
+      m_unsafe_ptr(*valobj_sp.get()) {
+
+  ProcessSP process_sp = valobj_sp->GetExecutionContextRef().GetProcessSP();
+  if (!process_sp)
+    return;
```
Why is this check here? `process_sp` doesn't seem used.

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,250 @@
+
+lldb::ValueObjectSP lldb_private::formatters::swift::
+    UnsafeBufferPointerSyntheticFrontEnd::GetChildAtIndex(size_t idx) {
+  if (idx > m_unsafe_ptr.GetCount())
+    return ValueObjectSP();
```
This check is redundant with the ones a few lines bellow. On top of that it is also bogus as a value of `num_children` would overflow.

fredriss commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -0,0 +1,250 @@
+  lldb::ProcessSP process_sp(valobj_sp->GetProcessSP());
+  if (!process_sp)
+    return false;
+  m_ptr_size = process_sp->GetAddressByteSize();
+  m_order = process_sp->GetByteOrder();
```
The pointer size and the byte order are never going to change. Set them in the constructor and never update them again. (That's actually a reason to get the process in the constructor).

fredriss reviewed (commented):
After you address this last round of comments, this LGTM

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test macOS Platform
